hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/DecayRpcScheduler.java

```diff
@@ -483,6 +501,12 @@ private void recomputeScheduleCache() {
 
     for (Map.Entry<Object, List<AtomicLong>> entry : callCosts.entrySet()) {
       Object id = entry.getKey();
+      // The priority for service users is always 0
+      if (isServiceUser((String)id)) {
```

Comment:
@tasanuma Thanks for your proposal. I am concerned any corner case here if put service users's request to the priority forever rather than flow controls. Such as user `hdfs` is one service user, and submit one big job which involve massive RPC request to NameNode, other normal users' request may postpone serious.

Reply:
Thanks for your comment, @Hexiaoqiao.
I think you can adjust the existing FairCallQueue settings to deal with that case to a certain extent. You could consider increasing the weight of the highest priority queue, or extending the length of the highest priority queue. However, this simple feature is not a feature that can be used in all cases. If more fairness is needed among service users, then I believe that [HADOOP-15016](https://issues.apache.org/jira/browse/HADOOP-15016) would be required.

The case where I think this feature would be useful is when you have service users who are regularly running critical ETL jobs. A user who can make huge RPC requests at a moment may not be suite for a service user.

Reply:
@tasanuma curious why you choose to do it here and `cachedOrComputedPriorityLevel` instead of `getPriorityLevel` in the previous patch.

Reply:
@sunchao Thanks for your review.
* When DecayRpcScheduler decays, it calls recomputeScheduleCache to recalculate the priority of all users, including service users, and put them in the cache. So we had to do this in recomputeScheduleCache as well. Lines 415-421 of the unit test cover this case.
* I moved it from getPriorityLevel to cachedOrComputedPriorityLevel because I thought it would be a little more efficient to determine if a user is a service user after looking at the cache information.

Reply:
Gotcha. This looks most good for me. One minor nit is that the logic for handling service users now is separated into two places. Another way perhaps is to add this in `computePriorityLevel` which is called by `recomputeScheduleCache` and `cacheOrComputePriorityLevel`. We need to add an extra parameter for this method though. What do you think?

Reply:
Thanks for your review, @sunchao! I agreed. Updated the PR addressing it.